File: Week 2/vigenere.c
/**
 * Name: Maria Daan
 * Student number: 11243406
 * Encrypts text into cyphertext by using a textual keyword
 */

#include <stdio.h>
#include <cs50.h>
#include <ctype.h>
#include <string.h>
#include <stdlib.h>

int main(int argc, string argv[])
{
    // Check if program is used correctly
    if (argc == 2)
    {
        string key = argv[1];
        int keylength = strlen(key);
        int nonalphacount = 0;

        // Check if a usable key has been implemented
        for (int j = 0; j < keylength; j++)
        {
            if (!isalpha(key[j % keylength]))
            {
                printf("Usage: ./vigenere k\n");
                return 1;
            }
        }

        // Get input from user: the text that needs to be encrypted and get ready to print the ciphertext
        string input = get_string("plaintext: ");
        int inputlength = strlen(input);
        printf("ciphertext: ");

        // Encrypt each letter from the input to an encrypted letter. Repeat the loop as many times as the length of the input
        for (int i = 0; i < inputlength; i++)
        {
            // Convert the key to upper case and make sure non-alpha characters do not effect which letter in the key is used to encrypt
            int upperkey = toupper(key[(i - nonalphacount) % keylength]);
            int alphkey = (upperkey - 65);

            // Make sure each lower case in the input encrypts and stays a lower case
            if (islower(input[i]))
            {
                int alphinput = (input[i] - 97);
                char cipher = ((alphinput + alphkey) % 26 + 97);
                printf("%c", cipher);
            }

            // Do the same for each upper case in the input
            else if (isupper(input[i]))
            {
                int alphinput = (input[i] - 65);
                char cipher = ((alphinput + alphkey) % 26 + 65);
                printf("%c", cipher);
            }

            // Keep track of spaces and print other characters than letters directly
            else if (!isalpha(input[i]))
            {
                nonalphacount++;
                printf("%c", input[i]);
            }
        }
        printf("\n");
    }

    // Make clear how the user is supposed to use the program when this hasn't been done right
    else
    {
        printf("Usage: ./vigenere k\n");
        return 1;
    }

    return 0;
}

File: Week 6/vigenere.py
# Problem Set 6
# Name: Maria Daan
# Student Number: 11243406
#
# This program encrypts messages using Vigenère’s cipher.

import sys
from cs50 import get_string

# check for (valid) key
if len(sys.argv) != 2 or not sys.argv[1].isalpha():
    print("Usage: ./vigenere key")
    exit(1)

# get/remember user input
key = sys.argv[1]
plaintext = get_string("plaintext: ")

ciphertext = ""
text_len = len(plaintext)
key_len = len(key)
alpha_len = 26
count = 0

# encrypt each letter from the plaintext to ciphertext
for i in range(text_len):

    # convert key to numeric value
    char_key = ord(key.upper()[(i - count) % key_len]) - ord('A')
    char_plain = plaintext[i]

    # make sure the plaintexts' case and non-alpha characters will be preserved
    if char_plain.isalpha():
        if char_plain.isupper():
            case = ord('A')
        else:
            case = ord('a')

        # convert plaintext to ciphertext by using the key
        cipher = (ord(char_plain) - case + char_key) % alpha_len + case
    else:
        count += 1
        cipher = ord(plaintext[i])

    ciphertext += chr(cipher)

print(f"ciphertext: {ciphertext}")